sdk/proto/provider.proto

```diff
@@ -112,6 +115,37 @@ message InvokeResponse {
     repeated CheckFailure failures = 2; // the failures if any arguments didn't pass verification.
 }
 
+message CallRequest {
+    // ArgumentDependencies describes the resources that a particular argument depends on.
+    message ArgumentDependencies {
+        repeated string urns = 1; // A list of URNs this argument depends on.
+    }
+
+    string tok = 1;                                        // the function token to invoke.
+    google.protobuf.Struct args = 2;                       // the arguments for the function invocation.
+    map<string, ArgumentDependencies> argDependencies = 3; // a map from argument keys to the dependencies of the argument.
+    string provider = 4;                                   // an optional reference to the provider to use for this invoke.
```

Comment:
Does this provider refer to the component provider? 

How does this interact with the provider map that is supplied on construct? How does a call impl retrieve that? 

https://github.com/pulumi/pulumi/blob/634e97cd55d8ad0347924b535f3f0daf6517a6c1/sdk/proto/provider.proto#L269

Reply:
Along the same lines, do resources created within call auto-parent to the component? Should this be configurable? 

Do we need to provide ways to expose all resource properties to call?

Reply:
Also config, and config secret keys.

Reply:
> Does this provider refer to the component provider?

No, it's the provider where the `Call` is implemented. This is the same as with `Invoke`. When making an `Invoke` or `Call`, you call it on the monitor, and you can include the provider/version to use:

https://github.com/pulumi/pulumi/blob/398fb2f8523e9d5c4fa0a064aaeab7ecdd70ee82/sdk/proto/resource.proto#L26

And the engine will lookup the specified provider (otherwise using the default provider) and forward the call:

https://github.com/pulumi/pulumi/blob/398fb2f8523e9d5c4fa0a064aaeab7ecdd70ee82/pkg/resource/deploy/source_eval.go#L559-L561

https://github.com/pulumi/pulumi/blob/398fb2f8523e9d5c4fa0a064aaeab7ecdd70ee82/sdk/proto/provider.proto#L37

> How does this interact with the provider map that is supplied on construct? How does a call impl retrieve that?

It's unrelated. There's not currently a way to fetch a component's map of providers outside of when the component is being registered. If there are particular providers that you want to use when creating resources from methods, you'll either have to have the component save them in the state, or pass them in as arguments.

> Along the same lines, do resources created within call auto-parent to the component? Should this be configurable?

No, there's no auto-parenting.

> Do we need to provide ways to expose all resource properties to call?

I'm not sure what you mean. For methods, there is a `__self__` argument that is a resource reference to the component, so when it is deserialized, it will have all of the component's state available for use.

> config secret keys.

Yep, I added this while rebasing.

Reply:
From a practical perspective:

```ts
const x = new Cluster("foo", { providers: { aws: explicitAwsProvider }});
const y  = x.AddSubnet();
```

I would find it very surprising as a user that resources created in `x.AddSubnet()` used the default AWS provider. The interaction of resources options provided to a component and subsequent usage of `Call` is something we should make an explicit design decision on. 

Personally, this seems like it will be common enough that telling users to transfer resource options into state in order for `Call` to have access is way too painful.

Reply:
> Personally, this seems like it will be common enough that telling users to transfer resource options into state in order for Call to have access is way too painful.

I agree. We should match the existing behavior to the greatest extent possible, which may mean exploring additional options for persistence. This might force the issue a bit on using a resource reference -> object table to ensure that any resource that has been marshaller in or out of a process is always represented by the same object (which would of course simplify other aspects of authoring as well).

Reply:
> This might force the issue a bit on using a resource reference -> object table to ensure that any resource that has been marshaller in or out of a process is always represented by the same object

Yes, this would definitely make things easier for implementers. Let me explore this a bit in the next day or two. If we're going to make a change along these lines, it'll be in a subsequent PR(s).